File: ios/Sources/Omnesis/Photos/Analyzers/BarcodeAnalyzer.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(Vision)
import Vision

/// Barcode/QR detection via `VNDetectBarcodesRequest` — new-photos tier
/// only. The decoded payload text is folded into `content` AND
/// promoted to `metadata.tags`; the raw payload + symbology are kept in
/// `metadata.extra` for rendering only, per the serving rule.
public struct BarcodeAnalyzer: PhotoAnalyzer {
    public let identifier = "barcode"

    public init() {}

    public func isAvailable() async -> Bool {
        true
    }

    public func analyze(_ input: PhotoAnalysisInput) async -> PhotoAnalysisFragment? {
        guard let image = input.image else { return nil }
        let request = VNDetectBarcodesRequest()
        let handler = VNImageRequestHandler(cgImage: image, options: [:])
        do {
            try handler.perform([request])
        } catch {
            return nil
        }
        let payloads = (request.results ?? []).compactMap(\.payloadStringValue)
        guard !payloads.isEmpty else { return nil }
        return PhotoAnalysisFragment(
            textLines: payloads,
            tags: payloads,
            extra: ["barcodePayloads": .array(payloads.map { JSONValue.string($0) })]
        )
    }
}
#endif

File: android/feature-photos/src/main/kotlin/dev/omnesis/android/feature/photos/analyzers/BarcodeAnalyzer.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.feature.photos.analyzers

import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.common.InputImage
import dev.omnesis.android.feature.photos.PhotoAnalysisFragment
import dev.omnesis.android.feature.photos.PhotoAnalysisInput
import dev.omnesis.android.feature.photos.PhotoAnalyzer
import kotlinx.coroutines.tasks.await
import kotlinx.serialization.json.JsonPrimitive

/**
 * Barcode/QR decoding via ML Kit's Barcode Scanning (already a dependency
 * elsewhere in the app, for QR device-pairing) — new-photos tier only. The
 * decoded payload text is folded into content AND promoted to
 * `metadata.tags`; the raw payload + symbology are kept in `metadata.extra`
 * for rendering only, per the serving rule. Mirrors iOS's `BarcodeAnalyzer`
 * (`VNDetectBarcodesRequest`).
 */
class BarcodeAnalyzer : PhotoAnalyzer {
    override val identifier = "barcode"

    private val scanner = BarcodeScanning.getClient()

    override suspend fun isAvailable(): Boolean = true

    override suspend fun analyze(input: PhotoAnalysisInput): PhotoAnalysisFragment? {
        val bitmap = input.bitmap ?: return null
        val barcodes = scanner.process(InputImage.fromBitmap(bitmap, 0)).await()
        val payloads = barcodes.mapNotNull { it.rawValue }.filter { it.isNotBlank() }
        if (payloads.isEmpty()) return null

        return PhotoAnalysisFragment(
            textLines = payloads,
            tags = payloads,
            extra = mapOf(
                "barcodeFormats" to JsonPrimitive(
                    barcodes.mapNotNull { it.rawValue?.let { _ -> it.format.toString() } }.joinToString(","),
                ),
            ),
        )
    }
}
